drivers: otp: add OTP driver subsystem with STM32 Ethernet MAC adress loading

GseoC opened the pull request:
Hello,

This P-R has dependencies on #96598 (includes it).

The STM32N6 boards embeds the BSEC peripheral that offers an interface to read/program One Time Programmable (OTP) fuses. One of the possible use-case for using OTP is fusing Ethernet MAC addresses at manufacturing level.
This way, every product embeds its own MAC address.

Add a fuse subsystem destined to welcome all fuse handling drivers.
Simply define read()/program() APIs as a starting point.

Then, plug the fuse API to the NVMEM subsystem so that the STM32 Ethernet driver can get its MAC address through it.

This use-case will be extended to STM32MPU boards afterwards.

GseoC committed:
drivers: Introduce new OTP subsystem

Introduce a new OTP subsystem to be able to interact with One Time
Programmable(OTP) memory. For now, add basic read()/program() APIs.

Add the otp_stm32_bsec driver as first user to be able to access
OTP fuses on some STM32 platforms.

File drivers/otp.h is inspired by drivers/eeprom.h as the basic
features are similar.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>

GseoC committed:
nvmem: Add OTP API support

OTP drivers are the interface to One Time Programmable memory, which
is NVMEM. Add the interface with the OTP API to the NVMEM one.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>

GseoC committed:
drivers: ethernet: stm32: handle MAC address with net_eth_mac_load()

This new API handles if the MAC address should be fetched from NVMEM,
is static or be randomly generated. Use it so that the driver can
fetch the MAC address from the OTP fuses, when possible.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>

GseoC committed:
dts: arm: st: n6: Add BSEC node for OTP handling

The STM32N6 has a BSEC peripheral for the OTP management, add it to the
SoC device tree file. Moreover, there are 4 OTP words dedicated for
Ethernet MAC addresses, describe them in the NVMEM layout.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>

GseoC committed:
boards: st: stm32n6570_dk: read the MAC address from the OTP fuses

Use the "mac_address0" NVMEM cell to be able to read the ethernet MAC
address from the OTP fuses.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>

pdgendt commented on include/zephyr/drivers/fuse.h:
Thanks for opening this PR! This is a great starting point to discuss in the Arch meeting group, I think.

I went with a different approach to use the `syscon` API in https://github.com/zephyrproject-rtos/zephyr/pull/98140, either works for me, but we'll need to come to resolution on that first.

pdgendt commented on include/zephyr/drivers/fuse.h:
```diff
@@ -0,0 +1,115 @@
+ * See @a fuse_program() for argument description
+ */
+typedef int (*fuse_api_program)(const struct device *dev, off_t offset, const void *data,
+				size_t len);
```
The driver _can_ be the provider already? Thanks @ZhaoxiangJin for your proposal too, let's bring it together in the Arch meeting.

decsny replied:
there are plenty of API in the world with the same function signatures, does that mean they all do the same things? no, of course not. so this block is wrong.

mathieuchopstm commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -14,6 +14,8 @@ menuconfig ETH_STM32_HAL
 	select ETH_DSA_SUPPORT_DEPRECATED
 	select PINCTRL
+	select FUSE
+	select NVMEM
```
Maybe `if SOC_SERIES_N6X`?

GseoC replied:
I reworked this bit

mathieuchopstm commented on drivers/fuse/fuse_bsec_stm32.c:
```diff
@@ -0,0 +1,161 @@
+	if (len == 0U) {
+		LOG_ERR("Read OTP with a len of 0");
+		return -EPERM;
+	}
```
This should probably be brushed off with a `return 0;` instead?

GseoC replied:
It's rather a question of: do we want to support reading a length of 0. I think not. IMO it's fine as is but do not hesitate to argument if you think otherwise.

mathieuchopstm reviewed (commented):
Ignoring the architectural concern:

decsny commented on subsys/nvmem/nvmem.c:
```diff
@@ -40,5 +45,9 @@ int nvmem_cell_write(const struct nvmem_cell *cell, const void *buf, off_t off,
 
+	if (IS_ENABLED(CONFIG_NVMEM_FUSE) && DEVICE_API_IS(fuse, cell->dev)) {
+		return fuse_program(cell->dev, cell->offset + off, buf, len);
+	}
```
@pdgendt and @ZhaoxiangJin were debating about this on another PR. I think the question to be answered first is what is the problem space we are trying to solve and what requirements are there for a solution to those problems. Because otherwise I don't know how to compare the two ideas.

ZhaoxiangJin replied:
I will also organize my proposal and provide solutions to the issue raised by [Fin(maass-hamburg)](https://github.com/zephyrproject-rtos/zephyr/pull/98149#discussion_r2455728462) and [Erwan(erwango)](https://github.com/zephyrproject-rtos/zephyr/pull/98149#discussion_r2455760754)

pdgendt commented:
> Hi @pdgendt , Great news! On my side, I got very sick so I wont be able to get something ready for this week. I hope I can cook something for next week (mid-end).

Sorry to hear that, get well soon!

> If that's urgent, I dont mind someone picking my patches and reworking them. Otherwise, I'll do it soon.

No worries, I don't think there's any rush.

mathieuchopstm commented on include/zephyr/drivers/otp.h:
```diff
@@ -0,0 +1,128 @@
+ * @brief Interfaces for One Time Programmable (OTP) Memory.
+ * @defgroup otp_interface One Time Programmable Memory
+ * @since 4.4
+ * @version 1.0.0
```
API should be experimental at first? :slightly_smiling_face: 
```suggestion
 * @version 0.1.0
```

pdgendt reviewed (commented):
Thanks for the update, I would have preferred a separate PR for the OTP driver, but if others are fine combining with the NVMEM changes, I'm fine too.

For the OTP driver, it would be nice to have tests/docs/release notes/... (hence the separate PR)

For the NVMEM addition, I'd prefer a test case being added to `tests/subsys/nvmem/api/` with some simulated driver.

GseoC commented on include/zephyr/drivers/otp.h:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	return api->read(dev, offset, data, len);
+}
```
What do we expect from a zero length read? Should it really be up to the driver to handle this case? I'm not sure if this can be silently treated as a valid parameter by returning 0 in the driver code because nothing will be stored in `data`.

No strong opinion but it seemed strange to me.

GseoC commented on include/zephyr/drivers/otp.h:
```diff
@@ -0,0 +1,128 @@
+
+	if (api->program == NULL) {
+		return -ENOTSUP;
+	}
```
I don't know if every OTP-related technology allow programming of its memory depending on the life cycle. I guess it's fine if optional. I can change to  `-ENOSYS` if it's the typical value

GseoC commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -14,6 +14,9 @@ menuconfig ETH_STM32_HAL
 	select ETH_DSA_SUPPORT_DEPRECATED
 	select PINCTRL
+	select OTP
+	select OTP_BSEC_STM32 if SOC_SERIES_STM32N6X
```
Yes, good suggestion, thank you.

GseoC commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -14,6 +14,9 @@ menuconfig ETH_STM32_HAL
 	select PINCTRL
+	select OTP
+	select OTP_BSEC_STM32 if SOC_SERIES_STM32N6X
+	select NVMEM
```
Sorry, the sequential part is in the NET_ETH_MAC_DT_CONFIG_INIT() macro which evaluates the presence of the properties sequentially, IIUC

GseoC committed:
boards: stm32n6570_dk: enable OTP and BSEC support

On STM32N6x SoCs, secrets and persistent information such as Ethernet
MAC addresses are stored in the BSEC OTP fuses. Therefore, default enable
OTP support if ETH_STM32_HAL and NVMEM are enabled so that the MAC address
can be read from the BSEC OTP fuses through the NVMEM API.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>

GseoC committed:
tests: nvmem: Add OTP testcase

Add an overlay/config to test the NVMEM subsystem with an emulated
OTP driver.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>

GseoC committed:
drivers: otp: add OTP memory emulation driver

In order to make some basic tests on the OTP API, add an OTP memory
emulator driver. It implements the .program() and .read() APIs.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>

GseoC committed:
boards: native_sim: add OTP memory emulator node

Add an OTP memory emulator node to the board native_sim board device
tree file.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>

mathieuchopstm commented on drivers/otp/otp_bsec_stm32.c:
```diff
@@ -0,0 +1,155 @@
+	const struct bsec_stm32_config *config = dev->config;
+	BSEC_HandleTypeDef handle = { .Instance = config->base };
+	unsigned int base_otp = offset / BSEC_WORD_SIZE;
+	HAL_StatusTypeDef hal_ret = HAL_OK;
```
Nit: no need to initialize since it is assigned directly
```suggestion
	HAL_StatusTypeDef hal_ret;
```

mathieuchopstm commented on drivers/otp/otp_bsec_stm32.c:
```diff
@@ -0,0 +1,155 @@
+	const struct bsec_stm32_config *config = dev->config;
+	BSEC_HandleTypeDef handle = { .Instance = config->base };
+	unsigned int base_otp = offset / BSEC_WORD_SIZE;
+	HAL_StatusTypeDef hal_ret = HAL_OK;
```
```suggestion
	HAL_StatusTypeDef hal_ret;
```

mathieuchopstm commented on drivers/otp/otp_emulator.c:
```diff
@@ -0,0 +1,81 @@
+static K_MUTEX_DEFINE(otp_emul_lock);
+
+struct otp_emul_config {
+	uintptr_t base;
```
Since this is unused, maybe we can remove it?

aescolar commented on drivers/otp/otp_emulator.c:
```diff
@@ -0,0 +1,81 @@
+ */
+
+#include <nsi_cmdline.h>
+#include <soc.h>
```
Minor: not needed either

mathieuchopstm reviewed (commented):
Seems pretty good. Minor things here and there:

aescolar commented on drivers/otp/Kconfig.emu:
```diff
@@ -0,0 +1,9 @@
+	default y
+	depends on DT_HAS_ZEPHYR_OTP_EMUL_ENABLED
+	help
+	  Enable OTP emulator driver used for testing purposes.
```
maybe something like this:
```Suggestion
	  Enable OTP emulator driver used for testing purposes.
	  It will use an static RAM buffer per instance which size is defined
	  in the DT node.
```

aescolar commented on drivers/otp/otp_emulator.c:
```diff
@@ -0,0 +1,81 @@
+static uint8_t otp_emul_mem_##n[DT_INST_REG_SIZE(n)];						\
+												\
+static const struct otp_emul_config otp_emul_##n##_config = {					\
+	.base = (uintptr_t)DT_INST_REG_ADDR(n),							\
```
```suggestion
```

aescolar commented on drivers/otp/Kconfig:
```diff
@@ -0,0 +1,37 @@
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig OTP
+	bool "One Time Programmable (OTP) handling drivers [EXPERIMENTAL]"
```
Nit: (handling seems kinda redundant for a driver)
```suggestion
	bool "One Time Programmable (OTP) drivers [EXPERIMENTAL]"
```

GseoC reviewed (commented):
Oops, didn't post this :)

GseoC commented on dts/bindings/otp/zephyr,otp-emul.yaml:
```diff
@@ -0,0 +1,8 @@
+# Copyright 2020 Google LLC
```
I forgot to post my comment on this 3 days ago, which I did 1 hour ago. I change to ST for consistency but is your suggestion the preferred way?

erwango commented on boards/st/stm32n6570_dk/Kconfig.defconfig:
```diff
@@ -39,4 +39,14 @@ config VIDEO_BUFFER_USE_SHARED_MULTI_HEAP
 	default y
 endif # VIDEO
 
+if ETH_STM32_HAL && NVMEM
```
```
if NET_L2_ETHERNET && NVMEM
```

erwango reviewed (changes requested):
We should now check how to inform new users and avoid breaking existing users.

erwango commented on drivers/ethernet/eth_stm32_hal_common.c:
```diff
@@ -185,7 +150,12 @@ static int eth_initialize(const struct device *dev)
 
-	generate_mac(dev_data->mac_addr);
+	/* Expect random, static, or OTP-based MAC address */
+	ret = net_eth_mac_load(&cfg->mac_cfg, dev_data->mac_addr);
```
If not configured later by the API, current behavior is: 
- random + ST OUI if defined in dts, else
- local + ST OUI if defined in dts, esle
- hw info based + ST OUI

This PR should add NVEM defined as a first option, then default to the same behavior. So it should be like:
- NVMEM based if defined, else
- random + ST OUI if defined in dts, else
- local + ST OUI if defined in dts, esle
- hw info based + ST OUI

I'm open to change this, but then it should be done in a latter PR and make necessary global changes (like in tree boards configuration), advertise change in migration guide, ...

GseoC replied:
Currently, `net_eth_mac_load()` handles:

- random if defined in dts, else
- NVMEM based if defined, else
- local  if defined in dts, else
- error

This is not the previous behavior so you are suggesting this change before another change that changes the behavior?

```
	if (!IS_ENABLED(CONFIG_NVMEM)) {
#if defined(ETH_STM32_RANDOM_MAC)
		/* "zephyr,random-mac-address" is set, generate a random mac address */
		gen_random_mac(dev_data->mac_addr, ST_OUI_B0, ST_OUI_B1, ST_OUI_B2);
#else /* Use user defined mac address */
		dev_data->mac_addr[0] = ST_OUI_B0;
		dev_data->mac_addr[1] = ST_OUI_B1;
		dev_data->mac_addr[2] = ST_OUI_B2;
#if NODE_HAS_VALID_MAC_ADDR(DT_DRV_INST(0))
		dev_data->mac_addr[3] = NODE_MAC_ADDR_OCTET(DT_DRV_INST(0), 3);
		dev_data->mac_addr[4] = NODE_MAC_ADDR_OCTET(DT_DRV_INST(0), 4);
		dev_data->mac_addr[5] = NODE_MAC_ADDR_OCTET(DT_DRV_INST(0), 5);
#else
		uint8_t unique_device_ID_12_bytes[12];
		uint32_t result_mac_32_bits;

		/* Nothing defined by the user, use device id */
		hwinfo_get_device_id(unique_device_ID_12_bytes, 12);
		result_mac_32_bits = crc32_ieee((uint8_t *)unique_device_ID_12_bytes, 12);
		memcpy(&dev_data->mac_addr[3], &result_mac_32_bits, 3);

		/**
		 * Set MAC address locally administered bit (LAA) as this is not assigned by the
		 * manufacturer
		 */
		dev_data->mac_addr[0] |= 0x02;

#endif /* NODE_HAS_VALID_MAC_ADDR(DT_DRV_INST(0))) */
#endif
	} else {
		/* Expect random, static, or OTP-based MAC address */
		ret = net_eth_mac_load(&cfg->mac_cfg, dev_data->mac_addr);
		if (ret < 0) {
			LOG_ERR("Failed to load MAC (%d)", ret);
			return ret;
		}
	}
```

pdgendt replied:
> That can be done if preferred. It would containt the OTP subsystem without the BSEC driver + the emulation driver and the NVMEM test case?

Ideally the regular bits for a new driver class, usually:

- Documentation bits
- Maintainer area
- Test/sample
- Simulator/emulator
- Shell commands?

erwango replied:
> By the way, I cannot fin any reference to ETH_STM32_RANDOM_MAC. What is it?

Maybe you cleaned it ? https://github.com/zephyrproject-rtos/zephyr/blame/main/drivers/ethernet/eth_stm32_hal_common.c#L29

pdgendt commented:
Superseded by https://github.com/zephyrproject-rtos/zephyr/pull/101292 and https://github.com/zephyrproject-rtos/zephyr/pull/102403
